[DependenceAnalysis] Fix incorrect analysis of wrapping AddRec expressions

sebpop opened the pull request:
Fixes GitHub issue #148435 where {false,+,true} patterns reported
    "da analyze - none!" instead of correct "da analyze - output [*]!".

The issue occurs when AddRec expressions in narrow types create cyclic
patterns (e.g., {false,+,true} in i1 arithmetic: 0,1,0,1,0,1...) that
violate SIV analysis assumptions of linear, non-wrapping recurrences.

The fix detects potential wrapping by checking if step × iteration_count
exceeds the type's representable range, then classifies such expressions
as NonLinear for conservative analysis.

Add wrapping detection in checkSubscript() with fallback to exact and max
backedge taken count for variable bounds.

sebpop committed:
[DependenceAnalysis] Fix SIV test crash when no AddRec after propagation

Fixes GitHub issue #148435 where testSIV() would hit an assertion failure
when neither Src nor Dst expressions contain AddRec after constraint
propagation. This can occur when propagation simplifies expressions to
non-AddRec forms. The subscript is effectively a zero induction variable.

The fix falls back to ZIV analysis in this case, treating the simplified
expressions as loop-invariant, which is the correct behavior when all
induction variable references have been eliminated through propagation.
The patch also fixes a MIV case that may decay into a ZIV test.

Add missing NonLinear case to switch statements in propagation code
to prevent 'bad subscript classification' crash when subscripts are
reclassified as NonLinear after constraint propagation.

Add regression test to prevent future occurrences of this issue.

sebpop committed:
[DependenceAnalysis] Fix incorrect analysis of wrapping AddRec expressions

Fixes GitHub issue #148435 where {false,+,true} patterns reported
"da analyze - none!" instead of correct "da analyze - output [*]!".

The issue occurs when AddRec expressions in narrow types create cyclic
patterns (e.g., {false,+,true} in i1 arithmetic: 0,1,0,1,0,1...) that
violate SIV analysis assumptions of linear, non-wrapping recurrences.

The fix detects potential wrapping by checking if step × iteration_count
exceeds the type's representable range, then classifies such expressions
as NonLinear for conservative analysis.

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -6439,8 +6439,129 @@ void ScalarEvolution::setNoWrapFlags(SCEVAddRecExpr *AddRec,
 
-ConstantRange ScalarEvolution::
-getRangeForUnknownRecurrence(const SCEVUnknown *U) {
+std::optional<bool>
+ScalarEvolution::mayAddRecWrap(const SCEVAddRecExpr *AddRec) {
```
This function doesn't make sense. At the very least ,please read the semantics of `NoWrapFlags`  in the ScalarEvolution.h

sebpop replied:
> Why did you remove @nikic from the reviewers?

That's a bug in github's click-o-metric interface.  Not intended, but anyways  nobody stops you to add back nikic.

> This function doesn't make sense.
What "doesn't make sense" to you?
The name of the function or the contents?
Be more specific please in your review.

kasuga-fj replied:
As for SCEV side, it's basically as already pointed out. Regarding DA, as I mentioned in another reply, NW is insufficient.

kasuga-fj reviewed (changes requested):
Why did you remove @nikic from the reviewers? Even setting that aside, this patch seems full of mistakes.

nikic reviewed (commented):
(Only looked at the SCEV side of this.)

sebpop committed:
[DA] Simplify runtime predicate collection and extend to all dependence tests

Previously, predicates were collected using a local `Assume` vector.
This patch:

1. Removes local `Assume` vector, uses class member `Assumptions` instead.
2. Adds a `getNonRedundantAssumptions()` helper for deduplication.
3. Extends predicate collection to all dependence tests.

sebpop committed:
[SCEV] Fix NSW flag propagation in getAddExpr

SCEV was losing NSW flags during AddRec operations, causing Dependence
Analysis to add unnecessary runtime assumptions for inbounds GEPs.

This patch fixes getAddExpr: when combining AddRecs from the same loop, preserve
compatible NSW/NUW flags from input AddRecs instead of always using FlagAnyWrap.

sebpop committed:
[SCEV] Fix NSW flag propagation in getMulExpr

SCEV was losing NSW flags during AddRec operations, causing Dependence
Analysis to add unnecessary runtime assumptions for inbounds GEPs.

This patch fixes getMulExpr: when multiplying AddRecs by constants, preserve NSW
flags that were explicitly requested from inbounds GEP operations. The overflow
check was too conservative and cleared flags even when they were explicitly
requested via OrigFlags.

sebpop committed:
[SCEV] Fix NSW flag propagation in getGEPExpr

SCEV was losing NSW flags during AddRec operations, causing Dependence Analysis
to add unnecessary runtime assumptions for inbounds GEPs.

This patch fixes getGEPExpr: inherit flags from index expressions when GEP has
no explicit flags, allowing NSW flags from AddRec indices to propagate to the
final GEP result.

This eliminates spurious runtime assumptions in DA for expressions like
{0,+,(4 * %N * %M)} derived from inbounds GEPs, allowing proper dependence
analysis without conservative runtime checks.

sebpop committed:
[LAA] Fix WAW dependency analysis with negative distances

Previously, LAA would incorrectly classify Write-After-Write dependencies
with negative distances as safe Forward dependencies, allowing inappropriate
vectorization of loops with bidirectional WAW dependencies.

The issue occurred in loops like:
  for(int i = 0; i < n; ++i) {
      A[(i+1)*4] = 10;  // First store
      A[i] = 100;       // Second store
  }

The dependence distance from first store to second store is negative:
{-16,+,-12}. However, this represents a bidirectional WAW dependency
that DependenceAnalysis would report as 'output [<>]!', indicating
dependency in both directions.

This patch fixes LAA to properly detect WAW dependencies with negative
distances as unsafe, preventing incorrect vectorization.

The fix adds a check specifically for Write-After-Write dependencies
before the general negative distance handling, ensuring they are
classified as Unknown (unsafe) rather than Forward (safe).

A proper fix, not implemented here because it would be a major rework of LAA,
is to implement bidirectional dependence checking that computes distances
in both directions and detects inconsistent direction vectors.

sebpop commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -6439,8 +6439,129 @@ void ScalarEvolution::setNoWrapFlags(SCEVAddRecExpr *AddRec,
+    LLVM_DEBUG(dbgs() << "\t\tAdvanced constant range analysis proves no-wrap: "
+                      << *AddRec << "\n");
+    return false;
+  }
```
There are at least 3 places in SCEV arith ops that do not preserve the wrap flags. Fixed in https://github.com/llvm/llvm-project/pull/155145

With these patches there are fewer runtime assumptions collected in DA tests.

kasuga-fj commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -6439,8 +6439,129 @@ void ScalarEvolution::setNoWrapFlags(SCEVAddRecExpr *AddRec,
+
+  // Step 1: Check existing no-wrap flags from SCEV construction.
+  if (AddRec->hasNoSelfWrap() || AddRec->hasNoUnsignedWrap() ||
+      AddRec->hasNoSignedWrap()) {
```
NW is insufficient. DA handles affine format expressions and expects their "monotonicity". Things are getting complicated because signed and unsigned interpretations are mixed up in DA.

> Overall I'm left confused which property you actually need for DA. Is it nw, nsw, nuw, nusw or some combination thereof?

I don't think this question is simple enough to be answered easily. The only thing now I'm certain of is that "NW is not sufficient".

kasuga-fj reviewed (commented):
Even if the changes on SCEV side were correct, I think DA side would still be incorrect. As I mentioned in another reply, NW is insufficient.

I believe https://github.com/llvm/llvm-project/pull/154527 is the right direction, although it's only a first step towards addressing the underlying issue -- DA doesn't account for SCEV's wrapping behavior.

sebpop commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -1337,6 +1337,12 @@ class ScalarEvolution {
   /// sharpen it.
   LLVM_ABI void setNoWrapFlags(SCEVAddRecExpr *AddRec, SCEV::NoWrapFlags Flags);
 
+  /// Check if this AddRec expression may wrap, making it non-affine.
```
@kasuga-fj, I don't understand your message here.
Do you mean replacing "making AddRec non-affine" with something else?

I did drop ", making it non-affine" in my local git and pushed the new version for review.
